#145028639 Fix account tests

Andretalik opened the pull request:
#### What does this PR do?
- Adds unit tests to the accounts tests.
#### Description of Task to be completed?
- Add assertions to the existing tests.
#### How should this be manually tested?
- run the command: ```python manage.py test hc.accounts.tests```
#### What are the relevant pivotal tracker stories?
- #145028641 Fixing the accounts tests.

victoriaaoka committed:
[Feature #145028639] Add status code assertion in tests

victoriaaoka committed:
[Feature #145028639] Fix the test to assert the correct error code

Andretalik committed:
[Feature #145028639] Fix Test to Assert Creation of User

- Added the assertion that confirms that a user has been created and added to the database.
	:muscle:

victoriaaoka committed:
[Feature #145028639] Add assertion that Profile has been created

Andretalik committed:
[Feature #145028639] Fix the test to assert contents of the email.

- Fixed the assertion to confirm the contents of the email.
	:pencil2:

victoriaaoka committed:
[Feature #145028639] Add assertion for the contents of the response.

Andretalik committed:
[Feature #145028639] Removed commented out code

victoriaaoka committed:
[Feature #145028639] Add assertion for the contents of the response. change content to html format

collinmutembei commented on hc/accounts/tests/test_login.py:
```diff
@@ -18,14 +17,14 @@ def test_it_sends_link(self):
-
+        ### **Assert that a user was created
+        final_list = len(User.objects.all())
+        self.assertEqual(1, final_list)
```
While this asserts that the list queryset for users has one object in it, there is no telling if the object represents the user that was created. Add more statements to assert that.

collinmutembei commented on hc/accounts/tests/test_login.py:
```diff
@@ -18,14 +17,14 @@ def test_it_sends_link(self):
         self.assertEqual(mail.outbox[0].subject, 'Log in to healthchecks.io')
         ### Assert contents of the email body
-
+        self.assertIn("please open the link below:", mail.outbox[0].body)
```
This assertion checks for static content in the email body. Add at least one more to check for dynamic content in the email.

georgreen replied:
@collinmutembei 
This assert checks for the content inside the page HTML, meaning if in future or for some reason the content of the page is changed e.g the string changes to "Find a link to log in below", this test fails not because the functionality is broken but the page changed. Would it be the case to assert that the content was returned by checking the status code?

collinmutembei replied:
Good point @georgreen however, the correct status code is not sufficient on its own. That's why we need to assert the content as well...preferably the dynamic stuff unique to the every login email.

collinmutembei commented on hc/accounts/tests/test_switch_team.py:
```diff
@@ -11,21 +11,22 @@ def test_it_switches(self):
-
-        ### Assert the contents of r
+        r= self.client.get(url, follow=True)
+        self.assertIn("<title>My Checks - healthchecks.io</title>\n", r.content)
```
This assertion fails with the error below:
```python
TypeError: a bytes-like object is required, not 'str'
```
Fix this!

collinmutembei commented on hc/accounts/tests/test_switch_team.py:
```diff
@@ -11,21 +11,22 @@ def test_it_switches(self):
     def test_it_checks_team_membership(self):
         self.client.login(username="charlie@example.org", password="password")
 
         url = "/accounts/switch_team/%s/" % self.alice.username
```
The PR diff clearly shows that you are not the author of this line of code however, I recommend that you update this line and others that may be using `%` for string formatting to use `str.format()`.

Resource: [pyformat.info](https://pyformat.info/)

collinmutembei commented on hc/accounts/tests/test_team_access_middleware.py:
```diff
@@ -13,5 +13,7 @@ def test_it_handles_missing_profile(self):
         self.client.login(username="ned@example.org", password="password")
         r = self.client.get("/about/")
         self.assertEqual(r.status_code, 200)
+        number_of_users = len(User.objects.all())
```
Profile object, not user object.

collinmutembei reviewed (changes requested):
Attend to the things I have highlighted with comments and finish the remaining tests.

collinmutembei commented:
Also, update the title of this PR as per the guidelines in the [bestpractices](https://github.com/andela/bestpractices/wiki/Git-naming-conventions-and-best-practices#pr-naming) repo.

Andretalik committed:
[Feature #145028639] Converts the test content to bytes

Andretalik committed:
[Feature #145028639] Changed the output operator

victoriaaoka committed:
[feature 145028639] Fixes assertion for email content from static to dynamic

Andretalik committed:
[Feature #145028639] Placed markers to show comments that have already been dealt with

collinmutembei commented on hc/accounts/tests/test_login.py:
```diff
@@ -1,35 +1,40 @@
-        ### Assert that check is associated with the new user
+
+        # ## Assert contents of the email body
+        self.assertIn("\n\nhttp://localhost:8000/accounts/check_token/" + self.profile.token, mail.outbox[0].body)
```
What do you think is going to happen when this assertion is run on travis. `localhost`? port **8080**?

collinmutembei commented on hc/accounts/tests/test_login.py:
```diff
@@ -17,15 +17,15 @@ def test_it_sends_link(self):
         assert r.status_code == 302
 
-        # ## **Assert that a user was created
+        # Assert that a user was created #
```
Why add `#` at the end of the comment?

Andretalik replied:
I added the `#` at the end of the comment to mark the ones that we were tasked with as done.

collinmutembei replied:
The diff is sufficient to show the introduced code, which is the base criterion as to whether the task is done.

collinmutembei reviewed (changes requested):
Clearly, we need to have a session to walk through tests and some scenarios with their corresponding assert statements. In the meantime have a look at the comments I have made and [this](https://developer.mozilla.org/en-US/docs/Learn/Server-side/Django/Testing) resource on MDN.

Andretalik committed:
Removed the unneccesary `#` in the test code

Andretalik committed:
[Feature #145028639] Removed debugger from testcode

Andretalik committed:
[Feature #145028639] Fix indentation errors

Andretalik committed:
[Feature #145028639] Fix test

 - Added assertion for test_send_report

Andretalik committed:
[Feature #145028639] Fix test_login

 - Added the tests for assertion that the check is associated with the new user.

Andretalik committed:
[Feature #145028639] Removed the commented out code for debugging purposes

victoriaaoka committed:
[feature 145028639] Add assertion that API key is created

victoriaaoka committed:
[feature 145028639] Add assertion that API key is revoked

victoriaaoka committed:
[feature 145028639] Add test for redirection for multiple login.

Andretalik committed:
[Feature #145028639] Added tests on bad token login

 - Added the asseertion for the redirection when trying to login wiht a bad token.

victoriaaoka committed:
[feature 145028639] Changed implicit variable names to explicit

Andretalik committed:
[Feature #145028639] Deleted unneccessary swap file

victoriaaoka committed:
[feature 145028639] Add test for redirection url.

victoriaaoka committed:
Merge branch 'ft-fix-account-tests-145028639' of https://github.com/andela/Healthchecks_ThunkNalla into ft-fix-account-tests-145028639

victoriaaoka committed:
[feature 145028639] Add test for redirection url.

collinmutembei reviewed (changes requested):
Push the changes with the fixes as pointed out in the peer PR review.

collinmutembei commented:
Also, update the PR description as per our best practices.

victoriaaoka committed:
[feature 145028639] Assert redirects when wrong token is used.

victoriaaoka committed:
[feature 145028639] Assert user was created using count and check is related to a user.

victoriaaoka committed:
[feature 145028639] Assert Api key is created and is not null.

victoriaaoka committed:
[feature 145028639] Assert response is in html format

victoriaaoka committed:
[feature 145028639] Assert new profile object counts using count

victoriaaoka committed:
[feature 145028639] Correct check to pass build

collinmutembei commented on hc/accounts/tests/test_login.py:
```diff
@@ -1,37 +1,48 @@
+        check.save()
+        test_user = User.objects.get(email="alice@example.org")
+        check = Check.objects.get(id=test_user.id)
+        self.assertEqual(check.id, test_user.id)
```
This assertion is failing and causing the [build](https://travis-ci.org/andela/healthchecks-thunknalla/jobs/235994362) to fail because the variable `check` on line 37 is being overwritten on line 40. The latter is a queryset containing multiple Check objects created by Alice while the former is the check object that you ought to be asserting that it is associated with Alice. Delete line 40 update the assertion to be
```python
self.assertEqual(check.user.id, test_user.id)
```
Makes sense?

victoriaaoka replied:
Yes. I've made the changes.

victoriaaoka committed:
[feature 145028639] Correct check
